Add tutorial post for using JWT Authorizers on AWS APIs (#67)
auth0-blog/guest-writer: articles/2019-12-23-securing-aws-http-apis-with-jwt-authorizers.markdown

---
layout: post
title: "Securing AWS HTTP APIs with JWT Authorizers"
description: "A tutorial on integrating Auth0 with HTTP APIs built on AWS API Gateway"
date: "2019-12-23 08:30"
author:
  name: "John Brennan"

>   url: "jhnbrnn"

Reviewer: Is this a Twitter or Github account?

Author: Thanks for catching this - it's a GitHub account. I've updated accordingly.

  mail: "brennan.john@mac.com"
  avatar: "https://twitter.com/jhnbrnn/profile_image?size=original"
related:
- 2017-11-15-an-example-of-all-possible-elements

---

**TL;DR:** JSON Web Tokens (JWTs) can be used to authorize requests to HTTP APIs built on Amazon Web Services’ API Gateway. This tutorial will walk you through integrating Auth0 with such an HTTP API.

---
​
## An introduction to HTTP APIs in API Gateway
​

> AWS API Gateway offers several solutions for building scalable APIs. HTTP APIs ([link](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api.html)) are a simple form of API that makes proxy requests to either another HTTP endpoint or an AWS Lambda. Unlike API Gateway's REST APIs, HTTP APIs can’t leverage other AWS services, such as AWS Identity and Access Management (IAM), as _authorizers_: services that allow or restrict API access to clients based on criteria such as user, roles, IP addresses, and so on. 

Reviewer: For SEO purposes, you can make the keywords the link.

For example:

- HTTP APIs ([link](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api.html)) are a simple form of API  ...

becomes

-  [AWS HTTP APIs](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api.html) are a simple form of API  ...

> AWS API Gateway offers several solutions for building scalable APIs. HTTP APIs ([link](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api.html)) are a simple form of API that makes proxy requests to either another HTTP endpoint or an AWS Lambda. Unlike API Gateway's REST APIs, HTTP APIs can’t leverage other AWS services, such as AWS Identity and Access Management (IAM), as _authorizers_: services that allow or restrict API access to clients based on criteria such as user, roles, IP addresses, and so on. 

Author: 👍 I like it! Definitely clearer and parses more easily. I'll make the change.

> AWS API Gateway offers several solutions for building scalable APIs. HTTP APIs ([link](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api.html)) are a simple form of API that makes proxy requests to either another HTTP endpoint or an AWS Lambda. Unlike API Gateway's REST APIs, HTTP APIs can’t leverage other AWS services, such as AWS Identity and Access Management (IAM), as _authorizers_: services that allow or restrict API access to clients based on criteria such as user, roles, IP addresses, and so on. 

Author: That makes sense! I've reworded slightly to better explain the HTTP API concept in the next commit.

> AWS API Gateway offers several solutions for building scalable APIs. HTTP APIs ([link](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api.html)) are a simple form of API that makes proxy requests to either another HTTP endpoint or an AWS Lambda. Unlike API Gateway's REST APIs, HTTP APIs can’t leverage other AWS services, such as AWS Identity and Access Management (IAM), as _authorizers_: services that allow or restrict API access to clients based on criteria such as user, roles, IP addresses, and so on. 

Author: On reflection, based on my responses on the other comments about the reshaping that the opening paragraphs would take, I wonder if an additional intro paragraph - more relaxed, more targeted to draw people into the article - would be beneficial here?

> AWS API Gateway offers several solutions for building scalable APIs. HTTP APIs ([link](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api.html)) are a simple form of API that makes proxy requests to either another HTTP endpoint or an AWS Lambda. Unlike API Gateway's REST APIs, HTTP APIs can’t leverage other AWS services, such as AWS Identity and Access Management (IAM), as _authorizers_: services that allow or restrict API access to clients based on criteria such as user, roles, IP addresses, and so on. 

Author: Yep, that was the intent of this paragraph originally, but it's doing too much work currently to explain a lot of concepts. Separating _what_ HTTP APIs are from _why_ you'd use them (benefits, comparison to REST APIs, etc) would make sense.

> AWS API Gateway offers several solutions for building scalable APIs. HTTP APIs ([link](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api.html)) are a simple form of API that makes proxy requests to either another HTTP endpoint or an AWS Lambda. Unlike API Gateway's REST APIs, HTTP APIs can’t leverage other AWS services, such as AWS Identity and Access Management (IAM), as _authorizers_: services that allow or restrict API access to clients based on criteria such as user, roles, IP addresses, and so on. 

Reviewer: Definitely, the general tone of the content should be conversational but also the introduction should be as persuasive as possible. Essentially the reader should be sold on investing time reading the content. Usually one 1-2 paragraphs.

> HTTP APIs can, however, use JSON Web Tokens (JWTs) to provide access control using a **JWT Authorizer**.

Author: I responded to a comment further down that mostly covers this, but the short version is I agree that the introduction needs a clearer problem statement. My rough thought is that an effective way to frame it is also the simplest – the pain point is that you sometimes want to build APIs that aren't accessible to unauthorized users, and JWT Authorizers are the solution.

The _details_ of that problem then become a variant of what's already on the page - AWS has many solutions for authentication, but HTTP APIs can't use them (this is the complicating factor). This leads into the "value prop" section for external identity providers, which provide industry-standard authentication+authorization (via OIDC/OAuth2 specs) that HTTP APIs support out-of-the-box with minimal configuration.

> HTTP APIs can, however, use JSON Web Tokens (JWTs) to provide access control using a **JWT Authorizer**.

Reviewer: Sounds good. It could start on mentioning why developers would want to use HTTP APIs with AWS in the first place. Why this route?

> ## What is a JWT Authorizer?

Author: I haven't seen one in the AWS documentation so far, so a diagram would have to be created for the purposes of the article, I think.

> ## What is a JWT Authorizer?

Reviewer: Right, if there are no official diagrams, we sketch them by hand or use software for it. Diagrams truly help understand architectural concepts much better 👍

​

> A JWT Authorizer configures the endpoint of your HTTP API to expect a token, created by an Identity Provider such as Auth0, to be provided in HTTP requests to the API. The authorizer will validate the token and ensure the requesting client is allowed to access the endpoint. JWT Authorizers can also optionally restrict access based on authorization scopes in the JWT.

Reviewer: > The authorizer will validate the token 

You have capitalized **Authorized** in some cases but not in others. For consistency, I recommend using the capitalization throughout, please.

> The Authorizer will validate the token and ensure the requesting client is allowed to access the endpoint.

Author: 👍 Good catch, thanks. AWS' documentation is a little inconsistent but I tried to aim for capitalization in all cases here, though it looks like I missed a few.

> A JWT Authorizer configures the endpoint of your HTTP API to expect a token, created by an Identity Provider such as Auth0, to be provided in HTTP requests to the API. The authorizer will validate the token and ensure the requesting client is allowed to access the endpoint. JWT Authorizers can also optionally restrict access based on authorization scopes in the JWT.

Author: That makes sense! And it is an access token, so that concept could easily be included in a slightly expanded version of this section.

> A benefit of JWT Authorization is the ability to pair existing identity provider solutions with the low overhead of HTTP APIs. This provides security and authorization without needing to rely on heavier processes such as AWS IAM or resource policy files.

Author: I hear what you're saying - I'm thinking out loud a little, but I wonder if there needs to be another section in here to explain identity solutions (AWS IAM, resource policy files, external identity providers) and then lead into the JWT Authorizer explanation. It probably wouldn't be as simple as that, since there's overlap between the existing sections, but you're right that it feels like a solid value prop statement is missing currently. I'd have to give this some more thought but the beginning sections framing the problem could definitely use some revamping and reorganizing.

## What will you build?

> In this tutorial, you’ll learn how to create add a JWT Authorizer to an HTTP API, using Auth0 as the identity provider. ​You’ll be building an HTTP API on AWS using API Gateway, a Lambda function, and DynamoDB. The API will support both GET and POST requests and read/write data from a DynamoDB table. POST requests will be restricted to authorized users and GET requests will be unauthorized, meaning anyone can access them.

Reviewer: > you’ll learn how to create add a JWT Authorizer 

There is an extra verb in here, "create add".

> In this tutorial, you’ll learn how to create add a JWT Authorizer to an HTTP API, using Auth0 as the identity provider. ​You’ll be building an HTTP API on AWS using API Gateway, a Lambda function, and DynamoDB. The API will support both GET and POST requests and read/write data from a DynamoDB table. POST requests will be restricted to authorized users and GET requests will be unauthorized, meaning anyone can access them.

Author: That makes a lot of sense, and agreed that this section should have lower complexity.

> Because it's the holidays, your API will create and retrieve a wish list of gifts you'd like to be given! Anyone will be able to read the wish list, but only authorized users will be able to add items to the list.

Author: 👍 Makes sense, I'll remove and reword slightly.

## Prerequisites

> * Auth0 Account

Author: Thanks for the link! I've updated my Prerequisites section to be smoother and call out signing up better.

* AWS Account
* Knowledge of Node.js (for creating the Lambda)
	​
## Configure Auth0
### Create new Auth0 API

> Head to your Auth0 dashboard and click “APIs” in the left-hand menu. Click the “Create new API” button and fill out the form with the following values:

Author: That makes sense - I changed both the section above (to more explicitly call out signing up for Auth0) and changed the reference here to clarify as well.

**Name:** AWS JWT Authorizer
**Identifier:** https://auth0-jwt-authorizer.com
**Signing Algorithm:** RS256 (default value)

> Click “Create”, and you’ll be taken to the definition page for your new API.

Reviewer: ```suggestion
Click "Create," and you'll be taken to the definition page for your new API.
```

As strange as it may look, the comma has to come before the closing double quotation mark 😬 

Please apply this feedback in the other instances of this pattern. I suggest doing a search for `",` to find them and update them 👍 

This also applies to period: `".` to `."`

Author: Thanks for this clarification - I struggle with this one a bit, as it's grammatically correct, but when referencing button text in the article, I tend to favor correctness in quoting the text over the grammar rule. Happy to change it, though! I've also changed a couple of the quote uses to monospaced text to ensure correctness.

![an example of the Auth0 API settings](images/jwt1.png)

You’ll want to copy or write down your API’s identifier; we’ll need it later on.
​
## Create HTTP API (Wish List API) on AWS

For this step, we’ll be using three services in AWS:

* DynamoDB (a NoSQL database to store our wish list items)
* Lambda (the business logic that will power our API, implemented as a serverless function)
* API Gateway (to create the public API endpoint itself and add the JWT Authorizer)

### Create DynamoDB

Navigate to the [AWS DynamoDB dashboard](https://console.aws.amazon.com/dynamodb/home), signing into AWS if needed. Click the “Create Table” button to get started.

We’ll name our table “WishList” and give the Primary Key a name of “id”, leaving its type as the default (String). Leave the rest of the form as-is; just scroll down and click Create.

![Settings for the DynamoDB table](images/jwt2.png)

That’s all you need to do for setup, so it’s time to head over to Lambdas!

### Create Lambda

Navigate to the [Lambda dashboard](https://console.aws.amazon.com/lambda/home) and click the “Create function” button in the table’s top right corner.

> The Lambda service provides a lot of helpful blueprints and samples, but for this tutorial you’ll be creating your Lambda's functionality from the ground up, so leave “Author from scratch” selected at the top of the form.

Reviewer: ```suggestion
The Lambda service provides a lot of helpful blueprints and samples, but for this tutorial, you'll be creating your Lambda's functionality from the ground up, so leave "Author from scratch" selected at the top of the form.
```

Author: Quick question about this one - I'm happy to add the comma, but I also noticed you changed the quotation symbols (`“”`) with primes (`""`). Is that intentional? Is one favored over the other?

Reviewer: It boils down to consistency, might had been my keyboard doing it. We favor the usage of the straight ones because it's what the keyboard inputs and what's most commonly used. It's just a style preference 👍

Give your function the name “wish-list-service” and set the runtime to “Node 12.x”. In the Permissions section, open up the accordion for creating an execution role; for this tutorial, you’ll use a template for setting permissions to allow the Lambda to use the DynamoDB table you just created.

> For Execution Role, Select “Create a new role from AWS policy templates”. Name the role “wish-list-service-role” and, in the Policy templates dropdown, select “Simple microservice permissions”.

Author: It's a shortcut to allowing the Lambda to read/write to DynamoDB. I've added an explanation, plus a link to the IAM dashboard, to this section in the following commit.

![Settings for the Lambda](images/jwt3.png)

Click “Create Function” in the bottom right corner of the page, and you’ll shortly be taken to the details page for your new Lambda. And now, it's time to code!

Start by replacing any existing code with the following empty handler function:

```js
exports.handler = async (event, context) => {

};
```

> You’re using the `async` keyword because the code will asynchronously access DynamoDB, which you’ll see in a moment. `event` and `context` will both be used in a moment to retrieve information about the API request.

Author: 👍 I've fleshed out this section in the next commit to explain more about the arguments being passed in.

Next, you’ll want to set up the connection between your Lambda and DynamoDB to perform your reads and writes:

```js
// NEW: import the AWS JavaScript SDK and create a new DocumentClient
const AWS = require('aws-sdk');
const ddb = new AWS.DynamoDB.DocumentClient();

const TABLE_NAME = "WishList";

// Existing code
exports.handler = async (event, context) => {

}
```

> `DocumentClient` will allow you to perform actions against your DynamoDB tables. The AWS documentation for DocumentClient can be found [here](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/DynamoDB/DocumentClient.html).

Author: I ended up removing this sentence and linking on the keyword of `AWS.DynamoDB.DocumentClient()` in the previous sentence.

Now you're ready to write your handler method. Since you’re handling both GET and POST requests in one handler, a simple `switch` statement based on the HTTP method should do the trick:

```js
// ...

exports.handler = async (event, context) => {
	switch(event.httpMethod) {
		case "GET":
			break;
		case "POST:
			break;
		default:
			break;
	}
}

```

With the scaffolding in place, it’s time to implement the GET case. It’s simple enough: when a GET request is received, you’ll retrieve records from the `WishList` table and return them:

```js
// ...

> switch(event.httpMethod) {

Author: The Lambda editor does provide code formatting (under "Edit" -> "Code Formatting"; it uses JSBeautify by default) and it does allow custom formatters, but it's not a straightforward process. From what I can tell, you'd have to deploy the Lambda code as a ZIP file and bundle up its node_modules dependencies (e.g. prettier) - but at that point you're working locally on code and you wouldn't need your formatter on the web editor.

	case "GET":

> 		// scan() will retrieve all records found in the table matching the TableName parameter

Author: That makes sense, I can move the comments out. In general, is there a style preference to breaking down the code before or after the code block? My default would be to explain it after the code block, but I'm open to suggestions.

        }),
        headers: {
          "Content-Type": "application/json",
        }
      };
    default:
      break;
  }
};
```

After you’ve added your code, click the orange “Save” button in the top right corner. Your Lambda is now live, and you’re ready to create the HTTP API that will make this Lambda’s functionality publicly available.

### Create an HTTP API in front of your Lambda

There are a few different ways to create an HTTP API; for this tutorial, you’ll be creating the API from the Lambda details page in which you’ve been working. By creating it from the Lambda, AWS will configure the API’s permission settings by default, allowing it to execute the Lambda when a request is received.

Scroll to the top of the page and open the “Designer” section. Click the “Add trigger” button in the section’s left side. Select “API Gateway” in the form’s dropdown, then “Create a new API” from the next dropdown that appears. In the form elements that follow, HTTP API should be selected by default, and you don’t need to change any of the settings - just head to the bottom of the form and click “Add”.

![Settings for the HTTP API Trigger](images/jwt4.png)

> By creating the API via the triggers section, everything should automatically wire up with the Lambda, so let’s give it a test using cURL to make sure. From the Designer section of the Lambda details page, click on the new API Gateway trigger object. From the API Gateway section that appears, grab the URL of your API and open up a new terminal window.

Author: That makes a lot of sense, thanks for the suggestion. I'll reword this to more clearly describe the Designer section.

> By creating the API via the triggers section, everything should automatically wire up with the Lambda, so let’s give it a test using cURL to make sure. From the Designer section of the Lambda details page, click on the new API Gateway trigger object. From the API Gateway section that appears, grab the URL of your API and open up a new terminal window.

Author: Update: I've pushed up a new version that's hopefully a little clearer. I don't think I've 100% got it right yet but it's closer. I've also added a new screenshot for clarity.

First, make a POST request to your endpoint to create a new wish list item:

```bash
$ curl -H "Content-Type: application/json" \
  -X POST \
  -d '{"name": "Test Item", "description": "Test Description", "url": "https://www.amazon.com"}' \
  https://[SUBDOMAIN].amazonaws.com/default/wish-list-service
```

> You should get back a 201 response with a JSON payload containing an `id` key.

Reviewer: 🤔 I did get the object response but not the response code. Maybe the cURL command needs a flag to show that detail?

Author: I've added the `-i` flag here to make the status code visible.

Next, make a GET request to ensure your new item was created:

```bash
$ curl https://[DOMAIN].amazonaws.com/default/wish-list-service
```

> The returning JSON payload should contain a key called `Items`, the value of which is an array of wish list items. Your new item will appear in that array.

Reviewer: The `Items` array is an array of id values tho, it doesn't seem to show any other info on the items.

Author: Odd...when I ran the example, the payload that the API call returned included all of the values stored on each item. Not sure where the discrepancy is happening but I'll try to play around with it and sort it out.

> ### Add JWT Authorizer using Auth0 App

Reviewer: > Add JWT Authorizer using Auth0 App

Perhaps "API" instead of "App" is more fitting with the Dashboard UI and Auth0 terminology.

> Now for the final step - setting up your Auth0 API as a JWT Authorizer. Head over to the [API Gateway dashboard](https://console.aws.amazon.com/apigateway/main/apis) and click “wish-list-service-API” to open up the API’s details page.

Reviewer: `wish-list-service-API` is not showing for me in that page. It just shows a menu to create different APIs. However, on the Lamba's page, under the "API Gateway" section, I can click on the API name there to open that page. Is that the same?

Author: Interesting - I'm not sure why it isn't showing up on the API Gateway dashboard, but it is the same. For simplicity's sake maybe I'll stick with navigating from the Lambda page, since that's how the API was created in the first place.

> First, navigate to the “Routes” section from the left-hand menu. Currently, the API is configured to allow any type of request to the `wish-list-service` endpoint, so you’ll need to change that first. Click “ANY” under the “/wish-list-service” Route in the left column, then click “Edit” in the details card’s header on the right. In the dropdown, change “ANY” to “GET” and click save.

Reviewer: ```suggestion
First, navigate to the “Routes” section from the left-hand menu. Currently, the API is configured to allow any type of request to the `wish-list-service` endpoint, so you’ll need to change that first. Click “ANY” under the “/wish-list-service” route in the left column, then click “Edit” in the details card’s header on the right. In the dropdown, change “ANY” to “GET” and click save.
```

![Changing the route to GET only](images/jwt5.png)

> You’ve now restricted the existing behavior - allowing requests to execute the Lambda - to only work on GET requests.

Reviewer: ```suggestion
You’ve now restricted the existing behavior &mdash; allowing requests to execute the Lambda &mdash; to only work on GET requests.
```

We use `&mdash;` to properly render those dashes in the content.

> You’ve now restricted the existing behavior - allowing requests to execute the Lambda - to only work on GET requests.

Reviewer: I think it would be cool to tell the reader to issue a POST request here, as an exercise, and see it fail. They will get this as response:

```
{"message":"Internal Server Error"}
```

Author: Makes sense, I've added that in.

> Once you’re back on the Routes page, click “Create” in the left column header to create a new route. Select “POST” from the methods dropdown and set the URL to “/wish-list-service”, then click Save. 

Reviewer: > ....  “/wish-list-service”, then click Save

My button label is "Create".

Author: Corrected!

![Adding a POST method to the route](images/jwt6.png)

You’ll be redirected to the Routes page; click “POST” under the “/wish-list-service” route in the left column to open up the details card.

First, you’ll need to configure the route to call your Lambda. To do so, click “Attach Integration” near the bottom of the details card. This will route you to the Integrations section of the API; once there, select “wish-list-service” from the dropdown of existing integrations and click “Attach Integration”.

![Attaching Lambda integration to your API](images/jwt7.png)

> Your API now allows POST as well as GET requests to run the Lambda.

Reviewer: I will add another cURL here to tests that the previous failed POST request now works.

> Next, head to the “Authorization” section via the left-hand menu. In the left column, click “Post” under the “/wish-list-service” route, then click “Create and attach an authorizer”. Fill out the form with the following details:

Reviewer: ```suggestion
Next, head to the “Authorization” section via the left-hand menu. In the left column, click `POST` under the “/wish-list-service” route, then click “Create and attach an authorizer”. Fill out the form with the following details:
```

> **Name**: auth0

Reviewer: ```suggestion
**Name**: auth0

```

> **Identity source**: $request.header.Authorization (this will use the Authorization header to check for the JWT)

Reviewer: ```suggestion
**Identity source**: `$request.header.Authorization` (this will use the Authorization header to check for the JWT)

```

> **Issuer URL:** https://[YOUR-DOMAIN].auth0.com (this can be found from your JWT metadata endpoint: https://[YOUR-DOMAIN].auth0.com/.well-known/openid-configuration, in the `issuer` field.)

Reviewer: ```suggestion
**Issuer URL:** `https://[YOUR-DOMAIN].auth0.com` (this can be found from your JWT metadata endpoint: https://[YOUR-DOMAIN].auth0.com/.well-known/openid-configuration, in the `issuer` field.)

```

> **Issuer URL:** https://[YOUR-DOMAIN].auth0.com (this can be found from your JWT metadata endpoint: https://[YOUR-DOMAIN].auth0.com/.well-known/openid-configuration, in the `issuer` field.)

Reviewer: Where can readers find this domain value? They would need to value of the domain in order to visit the JWKS endpoint, right?

Author: I'm not sure I fully understand your question - in general I wasn't too sure on how to represent "go to your Auth0 page and grab the URL". For clarity I've changed it to indicate that they should go to the JWKS endpoint and get the `issuer` field value, but if that's not the right track, feel free to let me know.

> **Issuer URL:** https://[YOUR-DOMAIN].auth0.com (this can be found from your JWT metadata endpoint: https://[YOUR-DOMAIN].auth0.com/.well-known/openid-configuration, in the `issuer` field.)

Reviewer: This bit has a lot going on. It may be better to use #### headers instead of a list as it would be easier to break down what's happening here. The issue here is missing the forward slash, `/` you show in . your screenshot, which will break the authorization.

Author: Makes sense - I've changed these to h4s and updated to include the the `/`

**Audience**: your Auth0 API’s unique identifier from the “Create new Auth0 API” step. If you didn’t write it down, you can find it by going to your Auth0 dashboard and navigating to the APIs page in the lefthand menu; the identifier for your API is listed on the landing page.

![Auth0 API Settings](images/jwt8.png)

![JWT Authorizer settings](images/jwt9.png)

Click “Create and attach”, and you’re done! You should now see a green “JWT” pill next to the POST method on your endpoint in the left column of the “Authorization” page.

> From this page, you could add scopes if you were using Role-Based Access Control in your Auth0 API, but for this tutorial, you’re all set up. The only thing left is to test it out.

Reviewer: I recommend linking the keywords "Role-Based Access Control" to an Auth0 page or document on the subject.

Reviewer 2: Actually, scopes and RBAC are not the same... they are not used together like this. Scopes is an official parameter supported by (defined in) the OAuth 2.0 spec that allows client to specify what exactly they want to do (what specific kind of delegated authz they are looking for). RBAC, on the other hand, uses other claims on the JWTs we (Auth0) produce as part of the authorization process to let applications know what permissions the person involved in the authz/authn process has.

The way we (Auth0) implements and employs RBAC has no formal definition, and it is not supported by any spec (AFAIK, but for sure it is not part of the OAuth 2.0 family of specs).

Author: Thanks for the clarification, @brunokrebs. This definitely would need rewording in another draft to create clarity around scopes and JWT Authorizers – or would you recommend cutting this line in order to keep the focus on the basics of setting up the Authorizer?

## Test it out!

> First, you should ensure the Authorizer is working properly. In your terminal window, make another POST request to the endpoint, ensuring that the `-i` flag is included to provide more request information:

Reviewer: > ensuring that the `-i` flag is included to provide more request information:

This is the flag that was missing in the previous cURL commands.

Author: Good catch, I've added `-i` to all cURL requests and I'll provide this context a little earlier in the article.

```bash
curl -H "Content-Type: application/json" \
  -X POST \

>   -d '{"name": "Test Item 2", "description": "Test Description 2", "url": "https://www.ebay.com"}' \

Reviewer: Not a deal breaker, but let's keep the links to amazon.com :P

  -i \
  https://[SUBDOMAIN].amazonaws.com/default/wish-list-service
```

> You should receive a 401 response, the body of which includes a `message` key with a value of `Unauthorized`. Your API no longer allows unauthorized creation of wish list items!

Reviewer: Please fence response codes: `401`, `200`, etc.

Your GET request should still work correctly, allowing unauthorized requests. Make a GET request with `-i`:

```bash
$ curl -i https://[DOMAIN].amazonaws.com/default/wish-list-service
```

And you should see a 200 response.

### Get Access Token from Auth0 Dashboard

> The moment of truth is here! Head to your Auth0 API dashboard and navigate to the details page of the custom API you created back at the beginning of this tutorial. Go to the “Test” section and copy the `access_token` property from the example response. 

Reviewer: Please include the link to the Dashboard here once again.

> The moment of truth is here! Head to your Auth0 API dashboard and navigate to the details page of the custom API you created back at the beginning of this tutorial. Go to the “Test” section and copy the `access_token` property from the example response. 

Reviewer: > from the example response. 

Suggestion:

> from the Response section.

> The moment of truth is here! Head to your Auth0 API dashboard and navigate to the details page of the custom API you created back at the beginning of this tutorial. Go to the “Test” section and copy the `access_token` property from the example response. 

Reviewer: Do we copy the access token as a whole or just the value of the `access_token` property? Do we include the quotation marks or just the string?

Reviewer 2: ![image](https://user-images.githubusercontent.com/15347252/71542716-69f0e300-2948-11ea-9ad8-d675fd6a9c25.png)

You can instruct them to use the "copy token" button on the top right.

Author: 👍 I've updated the instructions to use the "copy token" button (and also borrowed your screenshot 😄 )

![The access_token property from Auth0 test section](images/jwt10.png)

We’ll be adding this to our cURL POST request as follows:

```bash
curl -H "Content-Type: application/json" \
  -H "Authorization: [ACCESS_TOKEN]" \
  -X POST \
  -d '{"name": "Test Item 2", "description": "Test Description 2", "url": "https://www.ebay.com"}' \
  -i \
  https://[SUBDOMAIN].amazonaws.com/default/wish-list-service
```

(One important thing to note: JWT Authorizers don’t require the `bearer` keyword in the Authorization header. The value of the header is simply the access token itself.)

> That request should return a 201 status code - which means you’ve successfully integrated Auth0 as a JWT Authorizer on an AWS HTTP API!

Reviewer: This works if the "Issuer URL" is fixed in the Authorization configuration of this endpoint.

## Summary

In this tutorial, you’ve built a functional API using AWS Lambda, DynamoDB, and API Gateway. You’ve then added a JWT Authorizer, restricting access for POST requests to your endpoint to users that have authenticated with an Auth0 application. 

From here, there are several next steps you could take. For example, you could expand the functional capabilities of your API in the Lambda itself, or you could refine access by adding authorization scopes within Auth0 and restricting your API endpoints or methods using role-based access control.

> The sample code used to run the lambda can be found here: https://github.com/brennanjohn/a0-aws-jwt

Reviewer: > The sample code used to run the lambda can be found here: https://github.com/brennanjohn/a0-aws-jwt

"lambda" should be capitalized.

> The sample code used to run the lambda can be found here: https://github.com/brennanjohn/a0-aws-jwt

Reviewer: The repo URL is not clickable.
